### PR TITLE
Batch reducer improvements

### PR DESCRIPTION
**Pull request by llimeht**
Improve documentation and usability of batch reducer, including:

* adding an on-disk cache of reduced data so that the reducer can be restarted and plots can be made without requiring all data to be rereduced.

* making the list of reduced runs prettier by setting column types to `int` where appropriate and replacing `NaN` with blank entries.

### CODE REVIEW

File: refnx/reduce/batchreduction.py [hunk 1]
```diff
@@ -115,8 +131,24 @@ def add(self, row, ds, name, fname, entry, update=True):
         runs = run_list(entry)
         for run in runs:
             self.run_cache[run] = idx
+
+        if self.persistent:
+            self.write_cache()
         return data
 
+    def delete_rows(self, row_numbers):
```
Comment (andyfaff) [inline refnx/reduce/batchreduction.py]: needs a docstring.

File: refnx/reduce/batchreduction.py [hunk 2]
```diff
@@ -223,12 +258,49 @@ def _summary_dataframe(self):
         """
         df = pd.DataFrame(columns=self[0].entry.axes)
         for i, entry in enumerate(self):
-            df.loc[i] = entry.entry
+            if entry is not None:
+                df.loc[i] = entry.entry
         return df
 
+    def write_cache(self, filename=None):
```
Comment (andyfaff) [inline refnx/reduce/batchreduction.py]: docstring

File: refnx/reduce/batchreduction.py [hunk 3]
```diff
@@ -223,12 +258,49 @@ def _summary_dataframe(self):
...
+        with open(self._cache_filename(filename), 'wb') as fh:
+            pickle.dump(self, fh)
+
+    def drop_cache(self, filename=None):
```
Comment (andyfaff) [inline refnx/reduce/batchreduction.py]: docstring?

File: refnx/reduce/batchreduction.py [hunk 4]
```diff
@@ -223,12 +258,49 @@ def _summary_dataframe(self):
...
+        os.remove(self._cache_filename(filename))
+
+    def load_cache(self, filename=None):
```
Comment (andyfaff) [inline refnx/reduce/batchreduction.py]: docstring?

File: refnx/reduce/batchreduction.py [hunk 5]
```diff
@@ -223,12 +258,49 @@ def _summary_dataframe(self):
...
+        try:
+            if os.path.getsize(self._cache_filename(filename)) == 0:
```
Comment (andyfaff) [inline refnx/reduce/batchreduction.py]: if not x:

File: refnx/reduce/test/test_batch_reduce.py [hunk 6]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
         assert_equal(len(self.cache.name_search('A')), 3)
         assert_equal(len(self.cache.name_search(r'A\d')), 2)
         assert_equal(len(self.cache.name_search('no such sample')), 0)
+
+    def test_str(self):
+        # smoke test the __str__ method
+        assert str(self.cache) != ''
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: please replace all plain `assert` by `assert_`

File: refnx/reduce/test/test_batch_reduce.py [hunk 7]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+
+    def test_persistence(self, tmpdir):
+        # test persistence of cache in a local file
+
+        # presence of this file would indicate a failure of
+        # the previous persistent=False
+        assert not os.path.exists(self.cache._default_persistent_cache)
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

File: refnx/reduce/test/test_batch_reduce.py [hunk 8]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+
+        # make a new cache that has persistence with a default name
+        cache = refnx.reduce.batchreduction.ReductionCache(persistent=True)
+        self._addentry(self.entries[0], dest=cache)
+        assert_equal(len(cache), 1)
+        assert cache._cache_filename() == self.cache._default_persistent_cache
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

File: refnx/reduce/test/test_batch_reduce.py [hunk 9]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+        assert os.path.exists(cache._cache_filename())
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

File: refnx/reduce/test/test_batch_reduce.py [hunk 10]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+
+        # check that the persistent cache has has been dropped
+        cache.drop_cache()
+        assert not os.path.exists(cache._cache_filename())
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

File: refnx/reduce/test/test_batch_reduce.py [hunk 11]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+
+        # check that the cache filename can be set
+        cachename = tmpdir.mkdir('test').join('redn-test.pickle').strpath
+
+        # make a new cache that has persistence with specified filename
+        cache = refnx.reduce.batchreduction.ReductionCache(
+            persistent=cachename)
+
+        assert cache._cache_filename() == cachename
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

File: refnx/reduce/test/test_batch_reduce.py [hunk 12]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+
+        self._addentry(self.entries[0], dest=cache)
+        assert_equal(len(cache), 1)
+
+        # check that the persistent cache has been written
+        assert os.path.exists(cache._cache_filename())
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

File: refnx/reduce/test/test_batch_reduce.py [hunk 13]
```diff
@@ -122,3 +130,46 @@ def test_name_search(self):
...
+        assert os.path.getsize(cache._cache_filename()) > 0
```
Comment (andyfaff) [inline refnx/reduce/test/test_batch_reduce.py]: assert_

**Comment by llimeht** [comment]
Updated (although `assert` gives much more useful error messages than `assert_` and so I found that I was changing them to `assert` for debugging purposes...)

**Pull request closed without merging**